### ISSUE TITLE
Orleans Telemetry APIs

### ISSUE THREAD

**Issue by galvesribeiro** [dotnet/orleans#992]
## Problem

Today is hard to measure performance live on Orleans due to the lack of metrics and KPIs being extracted at runtime from Orleans runtime, and neither from Grain code.

We have today [ILogConsumer](https://github.com/dotnet/orleans/blob/master/src/Orleans/Logging/ILogConsumer.cs) which is pretty handy when we talk about Logging and tracing but not when we talk about KPIs.

The idea here is to create a set of APIs that will be used to extract in real-time information from Grains, Requests, Silos, Cluster, Streams, and many other Orleans Runtime components in order to push it to a [APM](https://en.wikipedia.org/wiki/Application_performance_management) external tool, or any kind of persistent storage for further processing or real-time display pushed data on a dashboard like [PowerBI](http://powerbi.com).
## Proposed Solution

**Telemetry Provider**

Based on common APM frameworks, provide a plugable abstraction for a APM provider such as this interface:

```
public interface ITelemetryProvider
{
    Task TrackDependency(string dependencyName, string commandName, DateTimeOffset startTime, TimeSpan duration, bool success);
    Task TrackDependency(DependencyTelemetry telemetry);
    Task TrackEvent(EventTelemetry telemetry);
    Task TrackEvent(string eventName,IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null);
    Task TrackException(ExceptionTelemetry telemetry);
    Task TrackException(Exception exception, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null);
    Task TrackMetric(MetricTelemetry telemetry);
    Task TrackMetric(string name, double value, IDictionary<string, string> properties = null);
    Task TrackRequest(RequestTelemetry request);
    Task TrackRequest(string name, DateTimeOffset startTime, TimeSpan duration, string responseCode, bool success);
    Task TrackTrace(TraceTelemetry telemetry);
    Task TrackTrace(string message);
    Task TrackTrace(string message, SeverityLevel severityLevel);
    Task TrackTrace(string message, SeverityLevel severityLevel, IDictionary<string, string> properties);
    Task TrackTrace(string message, IDictionary<string, string> properties);

}
```

This interface covers most of the reportable telemetry types for most of the APMs tools I've seen in the marked. A provider can be an APM tool/product where Orleans should post the telemetry data like for example [NewRelic](http://newrelic.com), Azure Application Insights, App Dynamics, DataDogHQ, etc., platform performance counters (on Windows regular performance counters, on Linux its own logd etc), a live even ingestion system like Azure EventHubs, or just a cold storage for metrics like a SQL Database, DocumentDb, Azure Table storage, whatever. The idea with the provider is to receive the telemetry incoming from Orleans clusters and store/process the way it usually does it.

The intention with the provider is to **report** telemetry data and not to analyze and neither parse telemetry. It behaves just as a proxy from the data coming from Orleans and the underlying APM Tool/storage APIs.

**Provider registration/configuration**

The providers should be registered using reflection and loaded at Silo startup. The configuration should be made by regular Orleans XML config, or by Orleans configuration objects in order to keep the config experience seamless with the other system components like Storage Providers for example. There can be multiple `ITelemetryProvider` registrations so all should post the telemetry to its underlying implementations in parallel e.g., you want the data to go to Event Viewer, Performance Counter and a SQL Database for different kinds of analysis. 

**Telemetry Consumer - System Grain (the "Consumer")**

This grain behave as a consumer from a internal stream where all the telemetry reported by Orleans is pushed from the producer (more on that later) and will forward the telemetry request to all the registered `ITelemetryProvider` providers in order push the telemetry forward to the underlying APM Tool/storage.

**Telemetry Manager - Proxy (the "Producer")**

One concrete **internal** class that behaves as a proxy for the telemetry providers. This class should implement the `ITelemetryProvider` interface in order to expose the same methods to be consumed by Orleans Runtime components so key measurable/critical points on Orleans infrastructure can be measured as KPIs.

This Proxy is responsible for receive the Telemetry and push this request to the stream in a **Fire & Forget** async model so there is no wait time during the regular process of a Grain method call (for example), since it is not a critical component for the actual request and should not impact its behavior. This class should be accessible internally by all Orleans infrastructure and from the `Grain`-based classes as well (more on that later).

**Custom metrics? Oh yeah!**

Most of the APM tools besides regular/common performance counters like requests p/ second, CPU, Memory, I/O, etc., also gives its users the ability to push their own technical and business metrics for further analysis, we should provide that ability for Orleans developers as well! 

The implementation of the _Telemetry Manager_ is internal to Orleans Runtime however, `ITelemetryProvider` is publicly available for Grain developers from the `Grain` class. This allow them to push custom telemetry just the same way as we do internally within Orleans framework boundaries. 

```
public class MyGrain : Grain, ISomeGrain 
{
  protected override Task<object> Invoke(InvokeMethodRequest request, IGrainMethodInvoker invoker)
  {
        base.TelemetryManager.TrackTrace("Hey just pushed a trace!");

        //Terrible sample on how measure request time but you guys got the idea :P
        var now = DateTime.UtcNow;
        await base.Invoke(request, invoker);
        var elapsedTime = DateTime.UtcNow.Subtract(now).TotalMiliseconds;
        base.TelemetryManager.TrackMetric("elapsedTime", elapsedTime);
        return Task.Done;
  }

  public Task MyGrainMethod()
  {
        base.TelemetryManager.TrackEvent("MethodCalled");
  }
}
```

Please add feedback on it. I would like to move forward with this issue since I saw many questions on gitter regarding telemetry, application insights and this is personally, one of my main pain points on Orleans since besides my business/custom KPIs, I have no vision on what is going on with my cluster. This could even enhance awesome ideas like [SiloMon](http://www.silomon.net) from @richorama and so many other monitoring tools that would be nice to tap on Orleans like _Microsoft System Center Operations Manager_ and so on.

Thanks

**Comment by johnkattenhorn** [comment]
This aligned very nicely with the patterns and standards we use and already do something similar within the grain code itself. 

We use Serilog and use it's sink adapter architecture to write out to NewRelic and Seq

**Comment by ashkan-saeedi-mazdeh** [comment]
You've thought it through very well.
Well the only issue as I see if I understand your design correctly is to rely on a single grain for it, The grain might get very busy and cause latency because its turns are taking much time. If we spilit the work between x number of grains per silo then the delay of this work being done will be divided between multiple timespans and will make the application more responsive. Imagine if the grain wants to send the results to the stream and there is really a lot of custom events from the user. Or we can have self imposed limits on the grain.

**Comment by galvesribeiro** [comment]
Perfect point @ashkan-saeedi-mazdeh !

When I mention a System Grain, I mean that this grain is internal to Orleans Runtime but it is mostly a stateless worker. They are only there to get the messages from the Stream and push to the underlying Telemetry Provider. Since it is a stateless worker, the instances will be created by Orleans Runtime under pressure. The idea of have a Fire & Forget model is to avoid latency on grain calls. 

Does that suppress your concerns or maybe I missed something? 

Thanks

**Comment by jason-bragg** [comment]
>"Today is hard to measure performance live on Orleans due to the lack of metrics and KPIs being extracted at runtime from Orleans runtime, and neither from Grain code."

Agreed, so you're suggesting we:
* Provide a PKI abstraction layer that can wrap a significant range of existing APMs.
* Enable Orleans framework and applications to use this PKI abstraction layer to expose performance metrics.

Is that correct?

>"Based on common APM frameworks"

Can you provide some links to the APIs of APM frameworks we may want to support or at least consider for design purposes?
Is there an existing APM implementation(s) that you'd suggest we initially support, or would the default implementation simply write to ILogConsumer? 

The existing proposal includes the programming model as well as implementation details (like grains and streams).  Can we start with a programming model first?
* What is the API grains and Orleans systems would use to record performance metrics? Without this it's hard to assess whether the ITelemetryProvider interface makes sense.

**Comment by galvesribeiro** [comment]
Ok thanks @jason-bragg let me try elaborate more on that...

> Agreed, so you're suggesting we:

> •Provide a PKI abstraction layer that can wrap a significant range of existing APMs.

> •Enable Orleans framework and applications to use this PKI abstraction layer to expose performance metrics.

Yes. By expose I meant to push it to the right APM tool blindly, without analyze it.


> Can you provide some links to the APIs of APM frameworks we may want to support or at least consider for design purposes?

I was think the first TelemetryProvider should be MS Application Insights + Performance Counters, then followed by New Relic after it.

The performance counters can replace the current Orleans.PerformanceCounters and AI is the easiest/cheapest way to start with APM for everyone.

> Can you provide some links to the APIs of APM frameworks we may want to support or at least consider for design purposes?

Sure! The sample interface for the `ITelemetryProvider` is nothing more than a base copy from Application Insights TelemetryClient which is already a terrific work from AI team to match similar APIs like NewRelic, DataDog and other big ones already in the market. Using a implementation of it for every APM provider, would keep things plugable as they are on Storage Providers API. With that in mind, we can replace ILogConsumer with the new APM API as Orleans.PerformanceCounters as well.

Some sample APIs:


[NewRelic - RecordMetric()](https://docs.newrelic.com/docs/agents/net-agent/features/net-agent-api#record_metric "NewRelic - RecordMetric()")

[DataDogHQ - Samples](https://github.com/DataDog/dogstatsd-csharp-client "DataDogHQ - Samples")

[Microsoft Application Insights](https://azure.microsoft.com/en-us/documentation/articles/app-insights-api-custom-events-metrics/ "Microsoft Application Insights")


There are a lot others but I believe this AI-inspired interface will suppress most of the APM tools on the market since they reporting SDKs talk pretty much the *same language*. 

>  Can we start with a programming model first?

The programming model for it is really simple. 

The Grain developer that which to consume it, just have to call inside a `Grain`-based class `base.TelemetryManager.TrackXXX(params);` and track what they want to do. Several pre-defined KPIs from Orleans systems will be already exposed and the developer can just push a new one anywhere they want to from a `Grain` class.

> What is the API grains and Orleans systems would use to record performance metrics? Without this it's hard to assess whether the ITelemetryProvider interface makes sense.

`ITelemetryProvider` has 2 purposes:

1. Be the contract for all APM tool provider implementation (NewRelic, AI, Performance Counters, etc)
2. Be the contract where the Grain developer can call methods in order to report custom metrics to all registered Telemetry Providers.

I think it is not that complicated to implement and neither to use. The grain developer will just have a new property on `Grain` class to call (`base.TelemetryManager`) its report methods. 

Does that make sense?

**Comment by jason-bragg** [comment]
Perfect.  Thanks!
I'll look into those links and get back with questions/comments.

**Comment by sergeybykov** [comment]
A good feature, no doubt. Let's see that we don't throw a baby with the bath water.

> ITelemetryProvider  has 2 purposes:
1.Be the contract for all APM tool provider implementation (NewRelic, AI, Performance Counters, etc)
2.Be the contract where the Grain developer can call methods in order to report custom metrics to all registered Telemetry Providers.

I agree with 1. I'm not convinced about 2. I would argue that the two contracts can and probably should be different. For example, I think that `Logger` has a better usability API for developers than `ITelemetryProvider.TrackTrace()`.

We could keep the existing `Logger` API as is and either extend it with a metrics reporting API or add a separate interface for metrics. Both of those can feed into a single `ITelemetryProvider` proxy for 1.

`ILogConsumer.Log()` also has extra handy information like exception, `IPEndpoint`, caller, log type that we benefit from in our current MDS/MDM-based monitoring because we can easily filter by those fields. It's unclear to me if `IDictionary<string, string> properties` is a sufficient replacement for those extra fields. Maybe. So my initial reaction is maybe let's leave `ILogConsumer` intact for now and add `ITelemetryProvider` as a parallel consumer for traces. We can always deprecate `ILogConsumer` later.

On the implementation side.
Do we have to call those telemetry consumers directly? Can't they consume data via ETW, so that the Orleans Runtime is completely decoupled from those monitoring technologies? That's how MDS/MDM works for us today - we install an implementation of `ILogConsumer` that dumps traces to ETW, and an MDS agent picks them up from there.

**Comment by galvesribeiro** [comment]
Thanks for the comments @sergeybykov and I hope it answers most of your doubts and the ones probably @jason-bragg will throw as soon as he get some time to read the APM tools samples I linked on my previous comment.

> I agree with 1. I'm not convinced about 2. I would argue that the two contracts can and probably should be different. For example, I think that  Logger  has a better usability API for developers than  ITelemetryProvider.TrackTrace() .

I believe that `TrackTrace` is not a happy name for sure. Like I mentioned before, it was mostly copied from *Application Insights* `TelemetryClient` class but, its open for a more reasonable name convention. If you take a look on some of the samples I've provided here from the APM tools, There are `RecordXXXXX()`, `PushXXXX()`, `TrackXXX()` but all them mean almost the same thing. No matter if it is a Log/Trace, Request, Metric, custom Event, Exception, etc., all them have internally a context which carry information about that telemetry and the environment (more on that later replying other part of your comment) and can be used for further analysis.

> We could keep the existing  Logger  API as is and either extend it with a metrics reporting API or add a separate interface for metrics. Both of those can feed into a single  ITelemetryProvider  proxy for 1.``

Ok, I think we can keep the `Logger` for now as a separate interface in order to not break existent code as a phase 1 for this change but, I don't think extend it with metrics is a good idea. 

Centralized Log systems/services like for example [Logentries](http://logentries.com "Logentries"), Event Viewer/ETW only deal with logs and basic events. They are hard to read, hard to search and have no correlation and context information at all. 

APM and Telemetry are another concepts. It can include logs and traces as well in order to provide additional information regarding a telemetry event for example, if an exception happens and is pushed to the APM tool, all the other telemetry events that happened near that can be correlated easily with some context information. Also, Telemetry in APM perspective is more focused on KPIs from the technical and business point of view in order to provide performance analysis on bother verticals.

If some developer don't care about the KPIs (like some people don't care about windows performance counters), but instead they want to just log/trace its application, they can just it with the same API and use a TelemetryProvider which will for example write to a table storage.

The hardest part today on get people moving from plain-old logging/etw/perfmon world into real APM (and the DevOps movement!) with telemetry(which also includes the 3 previously mentioned ways to monitor a system), is just a matter of culture and that they have to manually integrate new APIs into their codebase that are not "natural".

Just an analogy... How easy is to get data from a bunch of IIS logs and Performance Counters and put them into a dashboard? Pretty hard... Lots of manual parsing and handwork... Now think on do it in realtime and compare it to the same time on a time window from past time on historical data... Almost impossible(there still crazy dudes who do it today)! Well, APM is there for it, and if made in a easy way, I doubt people will keep looking at logs rather than look at correlated real-time telemetry :)

>  ILogConsumer.Log()  also has extra handy information like exception,  IPEndpoint , caller, log type that we benefit from in our current MDS/MDM-based monitoring because we can easily filter by those fields. It's unclear to me if  IDictionary<string, string> properties  is a sufficient replacement for those extra fields. Maybe. So my initial reaction is maybe let's leave  ILogConsumer  intact for now and add  ITelemetryProvider  as a parallel consumer for traces. We can always deprecate  ILogConsumer  later.


Lets get an example:

    Task TrackEvent(string eventName,IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null);
    
This first overload of `TrackEvent`(as any other `TrackXXX()`) is the simplest way to send an event from the developer point of view. He can just set the event name if this even is just to be later measured as a counter KPI, lets call a "*Checkout*" event on a e-commerce application. Now he knows how much Checkouts he had. Now, if he wants to enrich this even with its own business data, like for example a "*CustomerId*", he would add on the properties `IDictionnary<string, string>` which is optional. So now he wants to know the "*total order amount*" for that specific shopping cart, for that customer, when the Checkout event happened, so he just add "*TotalAmount*" to the metrics `IDictionnary<string, string>` which is also optional. Those are all properties that enrich the simple "*Checkout*" event and will be further analyzed. This a real scenario that I got from a customer that used to use MS Commerce Server before I left MS and it is spinned-off and this was reported as a performance counter! o.O

Well that is all about custom metrics for the business side, so what about the technical metrics from that? We can get it, and even better, we can correlate it so we know request time, IPEndpoints, caller, dependencies/external calls, if there is an exception, and put all together correlated with an APM tool and the right metrics. Now you know for each Checkout event, what are the technical conditions/events/traces that happened there. 

	Task TrackEvent(EventTelemetry telemetry);

Behind the scenes, the first overload of `TrackEvent()` just creates an `EventTelemetry` object (just a POCO class), add internal telemetry regarding that specific event (and its correlations if there is some) and then push it the registered APM `TelemetryProvider`. If the developer wants to create his own correlation (its common on all those APM tools) and discard the built-in KPIs from Orleans, just create a EventTelemetry object, fill its properties, and push up to the method. This same semantics of 1 automatic and 1 manual overload is there to all kinds of reports on the interface. Ofc the names of the methods are not written in stone so we can change it if there is some meaningful ones.

In other words, both properties and metrics dictionaries are there as optional parameters in order to give the developer the ability to enhance the telemetry along with the built in KPIs that came from Orleans runtime.

The idea is just to have a common interface to leverage a broad range of providers, not reinvent the well, and make this easy for developers using Orleans to plug in the APM tools that they love. Today, with ILogConsumer and Logger, they are locked on, well, logs, which would work fine if they just want it and want to push to some system like Logentries.


> Do we have to call those telemetry consumers directly

No, that's why I've mean to create the **Telemetry Manager - Proxy (the "Producer")** using which will implement `ITelemetryProvider` just as a facade proxy. The real provider is implementing this interface on an external DLL/Nuget and will added on the Silo Host project as developer desire (again, the idea is to allow multiple providers being registered since it is a common practice use more than one).

> Can't they consume data via ETW, so that the Orleans Runtime is completely decoupled from those monitoring technologies? That's how MDS/MDM works for us today - we install an implementation of  ILogConsumer  that dumps traces to ETW, and an MDS agent picks them up from there.

I don't think its a good idea. NewRelic and Application Insights has TraceListeners but, as their names implies, they just do log/trace. Telemetry still be caught by the application using their Nugets or posting directly to their APIs.

The proposed design is not to inject monitoring technologies or any kind of dependencies on Orleans runtime. The `ITelemetryProvider` is the only thing Orleans will have inside it (it will probably go to Core package). All the implementation is made by leveraging built-in features, mostly System Grains and internal Streams. 

The provider implementation, should be something like:



- Microsoft.Orleans.Telemetry.ApplicationInsights (Nuget Package)
	- ApplicationInsightsTelemetryProvider : ITelemetryProvider (The provider interface Implementation)
		- This is mostly a data mapper which will get info from properties on our objects (like `EventTelemetry` for example), map it to the configured target(s) `ITelemetryProvider`, and call their APIs.
	- Depends On:
		- Microsoft.Orleans.Core
		- Microsoft.ApplicationInsights 
    
So, **IF** the developer wants telemetry using a `TelemetryProvider`, they just have to configure it on `OrleansConfig.xml`:

    <TelemetryProviders>
		<TelemetryProvider type="Microsoft.Orleans.Telemetry.ApplicationInsights.ApplicationInsightsTelemetryProvider" ProviderKey="THE_APPLICATION_INSIGHTS_RESOURCE_API_KEY_HERE"/>
		
	</TelemetryProviders>

We can always enable some provider by default, like Performance Counters, trace file or Azure Table Storage so the developer will never notice the difference and don't have to do anything and their current Silos will already have this default data as they have today on Table storage for example.

The idea is to provide on v1 the following providers in order to maintain current non-APM users:


- Trace files 
- Console Tracer
- Debugger Tracer
- Azure Table Storage Trace events
- Azure Blob Storage Trace files/blobs
- Performance Counters
- Event Viewer

And, after this first core/built-in providers, add right after it the separated APM external tools  providers as ***Microsoft.Orleans.Telemetry.XXXX*** where XXXX is the APM product name:

- Azure Application Insights
- NewRelic
- DataDog
- Logentries
- Azure EventHub
- AppDynamics
- Any other which has community demand on to be implemented


If you prefer, we can still leave the current `ILogConsumer` and its infrastructure intact as you suggested and later on deprecate it if necessary. I just don't think its a big deal, since APM already cover logging and tracing while the opposite isn't true...


Does that make sense? I hope that answer your questions. Please feel free to shoot more! :)

**Comment by johnkattenhorn** [comment]
@galvesribeiro - Second your point about TraceListeners, telemetry has moved beyond simple textual messages which need to be read by some fragile parser (or a human), structured logging is much better. 

Right now in an Orleans project we use the TraceListener and re-direct that into Serilog which then have to parse it, before I go live I was going to look at a custom implementation of `ILogConsumer` to see if we could improve that situation to allow us to do some analysis on the log in NewRelic or Seq without having to parse the message directly.

For me the realization of this proposal couldn't come fast enough and I'd be happy to commit some effort even if touching the Orleans source code terrifies me :-)

**Comment by galvesribeiro** [comment]
@johnkattenhorn   yes... Reading logs today is, let's just say, a bad idea :)

I'm willing to implement this feature starting late night today across the weekend as soon as I have @sergeybykov's blessing. When the core is done, would be nice to have the community implementing more and more TelemetryProviders :)

**Comment by johnkattenhorn** [comment]
@galvesribeiro, count me in, I'll try to touchbase with you on gitter. It's time I did something for the project. 

I'm away this weekend but if you could do the implementation like you said then I'd happily build out telemetry providers, I have plenty of experience building Serilog Sink and using the NewRelic and AppInsights sdk.

**Comment by tsibelman** [comment]
Hi i very much like the direction this is going it looks like a very good addition,  but personaly i think splitting the telemetry provider into more modular  interfaces will be more convenient for future telemetry module implementers and users, metrics and logging can be implemented using different libraries for example.

**Comment by galvesribeiro** [comment]
@tsibelman  thanks for the feedback. 

I was thinking about it yesterday after my last reply to @sergeybykov's comment... 

There are APM products/services that has tracing and logging on it which is the case of NewRelic, Application Insights, DataDog so, have a different interface for those ones, will loose the correlation feature of it and will increase the configuration of the provider.

In other hand, services like Logentries doesn't have the telemetry side.

IMHO, it should be a single interface as it is for example on Microsoft System Center Operations Manager Management Packets. You have a single set of classes/interfaces. If your management pack collect Performance Counters, than you override/implement method X, otherwise, it has an empty implementation that does nothing. 

For most of the providers suggested here, they ALL collect telemetry except Logentries which still have the option to log a CustomEvent as a log line.

If there is a majority for split the interfaces in two (one for telemetry and other for logs) we can do that... I'm just not quite convinced yet that it is necessary for the majority of cases.

**Comment by tsibelman** [comment]
Not all packages provide metrics and logging together. For example right now orleans using perfcouners for metrics and tracelog for logging.  So If someone would like to stay with default  perf counters implementation but to provide a different logging framework it would be cumbersome to do so. But I definitely see the value in single interface from telemetry user perspective  a solution I can think right now is maybe to have an class that implements the original  interface and has two constructors one that has original interface as a parameter  and other that accepts two separated interfaces as parameters.  This way can have the flexibility to implement  telemetry in most apropiate way.

**Comment by galvesribeiro** [comment]
@tsibelman  I disagreen on that. Can you provide a sample class which demonstrate your suggestion? Maybe I'm missing something.

Regarding the perfcounters, it is a way to report telemetry while the trace on file or even event viewer is a way to make a logger. Those two implementations can be tied together on *ETW* concept since both, if we look at `TraceListener`, are being implemented on the same place, TraceListener provider so I see both as being a single `ETWTelemetryProvider` and on its XML config, it would have parameters like where to log is going to be written (file or EventViewer), some group configuration for EventViewer and perfcounter etc. 

Remember that the XML config isn't fixed to all providers. Each Telemetry provider has its own set of properties to being read from OrleansConfig.xml.

Also keep in mind that ETW scenarios are created for Windows (there is no such thing as PerfCounters on Linux systems) and we are going to move to coreCLR which is suppose to run on Linux and MacOS, so have a real abstraction of the programming model and data contracts in the simplest way as possible, would make easier to work with it on another systems.

Maybe I'm wrong and we should split it, right now I just don't see a point on that today.

Thanks

**Comment by jason-bragg** [comment]
I am of the opinion that there are two interfaces (or sets of interfaces) we need to consider.

**Client Interface**
This is the interface Orleans and Orleans applications have access to for KPI purposes.  I tend to lean towards @sergeybykov's suggestion on this, that since KPI's can include logging and performance counters it makes sense to extend the existing logging interfaces, rather than duplicating the current capabilities in a new api.

**APM Provider Interface**
This is the abstraction layer that allows developers to integrate APM's into Orleans.  As mentioned, since some APM's support logging, some support performance counters, and some support both, it make sense to have two interfaces here; A logging abstraction interface and a performance counter abstraction interface.  If a provider conforms to the logging interface, it will be wired up to the logging framework.  If it conforms to the performance counter interface, it will be wired up to the performance counters framework.  If it conforms to both, it will be wired up to both.  These checks can be made at provider load time.

With a framework like this, we can take the current logging capabilities and move them into a APM logging provider and the current PerfCounter logic and move it into an APM performance counter provider.  This would preserve existing behaviors for existing customers, while allowing the integration of APM's using @galvesribeiro's Microsoft.Orleans.Telemetry.XXXX  plan.

Thoughts?

**Comment by galvesribeiro** [comment]
> Client Interface

Ok I like your idea @jason-bragg. Looks like we have an agreement. As long as the developer has a single client interface to call logging and KPIs rather than different ones to deal with I'm up to a different client interface then the provider one as much as I think that we will be converging both interfaces (log and KPIs) methods on this client one, I think we are ok with that.

> APM Provider Interface

Regarding the separation of Logging and KPIs at the provider level, I'm also up for it, since it is a internal implementation and we can have a single provider class that implement both interfaces if that specific provider supports it. I like that.

So, if no one else has anything to add on that, now I'll only wait for @sergeybykov's green light so we can close this issue and start with the implementation as agreed here. I want to start late Tonight if possible in order to get it done this weekend and I'm whiling to make small pushes to a PR referencing this issue so we are all following the progress and changes can be made while items are going reviewed.

@sergeybykov its on you now :)

Thanks everyone for the feedback, this will be a nice feature IMHO to aggregate on Orleans framework.

**Comment by tsibelman** [comment]
Sounds good to me.  One additional question I have,  what would happen if multiple telemetry  providers are given in configuration are they chained somehow  or it a configuration error?

**Comment by jason-bragg** [comment]
I'm very happy to see how engaged you are on this (seriously, it's impressive), and I don't want to slow you down, but while you're working on the implementation, it would be beneficial if you posted the client and provider APIs at your earliest convenience.

Since Orleans is a framework the programming model is very important, and these interfaces will likely receive as much (or more) scrutiny as the implementation.  Unlike the implementation, if these interfaces need to change, that could break customer code.  By getting these interfaces out early their review may be complete before the implementation review.  Reviewing an implementation will be far smoother if the programing model has already been settled on.

**Comment by galvesribeiro** [comment]
@tsibelman No error. You can have as many providers as you want configured. They all will receive in parallel the telemetry/log request to be pushed to its underlying APM/Logger tools. Its up to the developer to decide how many they want to.

**Comment by galvesribeiro** [comment]
@jason-bragg  Orleans gave me awesome stuff to my company. Its nothing better than give back somehow something and also, working on something that is used on a game that I really like? Is something unbeatable.

So, regarding the interfaces for the client and the providers, I believe that it is just a matter of split the `ITelemetryProvider` in 2 interfaces `ILoggerTelemetryProvider` and `IKPITelemetryProvider` with their respective methods splitted to each other (names here are just a suggestion)... I think the client would be a single interface, ```ITelemetryProducer : ILoggerTelemetryProvider, IKPITelemetryProvider```

The client will have access to both methods from a single point, if there is a `ILoggerTelemetryProvider` provider registered when a developer calls `ITelemetryProducer.Log(xxxx)` then it is forwarded to the those providers, otherwise, nothing happens. 

The important thing is to have a single point of report for the developer. I'll build the interfaces and some data contracts and post here later today.

**Comment by jason-bragg** [comment]
@galvesribeiro Great!
Minor nit:  The term 'telemetry' is a very overused term that means very different things to different developers.  I'd personally avoid using it.  We learned this lesson well with our 'streaming' APIs :)

**Comment by galvesribeiro** [comment]
@jason-bragg thanks for the tip, was about to post the classes... What would you suggest? In my experience all APM tools use this term... Any idea?

**Comment by jason-bragg** [comment]
@galvesribeiro Oh I suck at naming things.  I've gotten better at recognizing names that could come back an bite me, but that's about.  Please don't let naming slow you down.

**Comment by galvesribeiro** [comment]
## Provider API ##

Those interfaces should be implemented by the APM/Logger providers respectively:

    public interface IKPITelemetryProvider
    {
        Task TrackDependency(string dependencyName, string commandName, DateTimeOffset startTime, TimeSpan duration, bool success);
        Task TrackDependency(DependencyTelemetry telemetry);
        Task TrackEvent(EventTelemetry telemetry);
        Task TrackEvent(string eventName, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null);
        Task TrackMetric(MetricTelemetry telemetry);
        Task TrackMetric(string name, double value, IDictionary<string, string> properties = null);
        Task TrackRequest(RequestTelemetry request);
        Task TrackRequest(string name, DateTimeOffset startTime, TimeSpan duration, string responseCode, bool success);
        Task TrackTrace(TraceTelemetry telemetry);
    }

    public interface ILoggerTelemetryProvider
    {
        Task TrackException(ExceptionTelemetry telemetry);
        Task TrackException(Exception exception, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null);
        Task TrackTrace(string message);
        Task TrackTrace(string message, SeverityLevel severityLevel);
        Task TrackTrace(string message, SeverityLevel severityLevel, IDictionary<string, string> properties);
        Task TrackTrace(string message, IDictionary<string, string> properties);
    }

## Client API ##

This interface will be implemented internally by Orleans Core and is the "frontend" for the APM framework as a single point to report both logs and KPIs.

    public interface ITelemetryProducer : IKPITelemetryProvider, ILoggerTelemetryProvider
    {
    }

When a developer want to send a log or a KPI to be dispatched for the registered matching providers, they would call something like this on a `Grain`-based class:

    base.TelemetryManager.TrackTrace("Hello APM world!");

## The contracts ##

Those are the classes that would be used to send telemetry to the providers. Those values would be mapped to the underlying provider classes/parameters in order to dispatch the requests to them. Note that those context classes are there to simplify the usage and is pretty much what is on Application Insights requests (again, great job for AI team!). We could add/remove fields as necessary to catch more information for the Cluster/request automatically.

    public class ClientContext
    {
        public string Id { get; set; }
        public string ClientName { get; set; }
        public LocationContext RemoteClientLocation { get; set; }
    }
     public class GrainContext
    {
        public string Id { get; set; }
        public string Reference { get; set; }
        public string Interface { get; set; }
        public string Type { get; set; }
    }
	public class InternalContext
    {
        public string OrleansVersion { get; set; }
    }
	public class InvocationContext
    {
        public GrainContext Grain { get; set; }
        public string MethodName { get; set; }
        public IList<string> Parameters { get; set; }
    }
	public class LocationContext
    {
        public string Endpoint { get; set; }
    }
	public class SessionContext
    {
        public string Id { get; set; }
    }
	public enum SeverityLevel : byte
    {
        Critical = 0,
        Error = 1,
        Information = 2,
        Verbose = 3,
        Warning = 4
    }
	public class SiloContext
    {
        public string Id { get; set; }
        public string OperationSystem { get; set; }
        public string RoleInstance { get; set; }
        public string RoleName { get; set; }
        public string Type { get; set; }
        public LocationContext EndPoint { get; set; }
    }
	public class TelemetryContext
    {
        public GrainContext Grain { get; set; }
        public InvocationContext Invocation { get; set; }
        public ClientContext Client { get; set; }
        public SessionContext Session { get; set; }
        public SiloContext Silo { get; set; }
        public InternalContext Orleans { get; set; }
        public IDictionary<string, string> Properties { get; set; }
    }

	public class DependencyTelemetry
    {
        public bool? Async { get; set; }
        public string Name { get; set; }
        public int? Count { get; set; }
        public string DependencyTypeName { get; set; }
        public TimeSpan Duration { get; set; }
        public string OperationName { get; set; }
        public IDictionary<string, string> Properties { get; set; }
        public DateTimeOffset StartTime { get; set; }
        public bool? Success { get; set; }
        public DateTimeOffset Timestamp { get; set; }
    }
	public class EventTelemetry
    {
        public TelemetryContext Context { get; set; }
        public IDictionary<string, double> Metrics { get; set; }
        public IDictionary<string, double> Properties { get; set; }
        public string Name { get; set; }
        public DateTimeOffset Timestamp { get; set; }
    }
	    public class ExceptionTelemetry
    {
        public TelemetryContext Context { get; set; }
        public Exception Exception { get; set; }
        public ExceptionHandledAt HandledAt { get; set; }
        public IDictionary<string, double> Metrics { get; set; }
        public IDictionary<string, double> Properties { get; set; }
        public SeverityLevel? SeveritLevel { get; set; }
        public DateTimeOffset Timestamp { get; set; }
    }

    public enum ExceptionHandledAt : byte
    {
        Platform = 0,
        Unhandled = 1,
        UserCode = 2
    }
	 public class MetricTelemetry
    {
        public TelemetryContext Context { get; set; }
        public int? Count { get; set; }
        public double? Min { get; set; }
        public double? Max { get; set; }
        public string Name { get; set; }
        public IDictionary<string, string> Properties { get; set; }
        public double? StandardDeviation { get; set; }
        public DateTimeOffset Timestamp { get; set; }
        public double Value { get; set; }
    }
	public class RequestTelemetry
    {
        public TelemetryContext Context { get; set; }
        public TimeSpan Duration { get; set; }
        public string Id { get; set; }
        public IDictionary<string, double> Metrics { get; set; }
        public IDictionary<string, double> Properties { get; set; }
        public string Name { get; set; }
        public DateTimeOffset StartTime { get; set; }
        public bool? Success { get; set; }
        public DateTimeOffset Timestamp { get; set; }
    }
	public class TraceTelemetry
    {
        public TelemetryContext Context { get; set; }
        public string Message { get; set; }
        public IDictionary<string, string> Properties { get; set; }
        public SeverityLevel? SeverityLevel { get; set; }
        public DateTimeOffset Timestamp { get; set; }
    }
	

PS: the XXXTelemetry-ish names we could change, just come with a suggestion and we are good :)

**Comment by jason-bragg** [comment]
There is a lot here, so it will take a little time to grock but I thought there was a general agreement that the client api would be an extension of the existing Orleans logging client.  (See Orleans.Runtime.Logger).

The Orleans client api will likely need some cleanup before being extended, but generally what we come up with should be a (mostly) natural extension of the existing code.

In this proposal, the client and provider interfaces are the same.  Even if the provider and client interfaces end up looking exactly the same, which I don't think they will, they should be different interfaces, because they perform different responsibilities and need to be independent.

For instance:
* ease of use is very important in the client interface, but the ability to support a wide range of APM's is probably more important in the provider interfaces.
* Things like SiloContext, GrainContext and the like should not be part of the client interfaces.  Application developers shouldn't be concerned with this level of detail.  The framework should fill in things like this and pass them through to the providers.
* I don't think the client interfaces can return tasks.  Applications can't be slowed down by asynchronous calls to record performance data.

Very rough and poorly thought through attempt at a client interface below.

    public interface ITelemetryLogger
    {
        // current logging stuff
        SeverityLevel Severity { get; }
        void Log(SeverityLevel severity, string message);
        void Log(SeverityLevel severity, int logCode, string message);
        void Log(SeverityLevel severity, int logCode, string message, Exception exception);
        // new telemetry logging stuff
        void Log(SeverityLevel severity, string message, IDictionary<string, string> properties, IDictionary<string, double> metrics = null);
        void Log(SeverityLevel severity, int logCode, string message, IDictionary<string, string> properties, IDictionary<string, double> metrics = null);
        void Log(SeverityLevel severity, int logCode, string message, Exception exception, IDictionary<string, string> properties, IDictionary<string, double> metrics = null);
        // new telemetry KPI stuff
        void TrackDependency(string dependencyName, string commandName, DateTimeOffset startTime, TimeSpan duration, bool success);
        void TrackEvent(string eventName, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null);
        void TrackMetric(string name, double value, IDictionary<string, string> properties = null);
        void TrackRequest(string name, DateTimeOffset startTime, TimeSpan duration, string responseCode, bool success);
    }

Not saying the above is right, but something like that could be integrated with the current Orleans.Runtime.Logger without breaking existing applications.

**Comment by sergeybykov** [comment]
@galvesribeiro First of all, I really appreciate you pushing this. This is a very exciting development of the telemetry story that has been stagnant for a long time. I am not pushing back on in in any way. I am just a bit obsessed with the developer facing APIs.

@jason-bragg Expressed most of my thought before I had a chance to. :-)

I'm of the even radical opinion that we can leave `Logger` as is with its what I consider developer friendly API for logging, and just add `TrackXXX` methods to it. We can turn it into an interface if needed. 

```csharp
public abstract class Logger
{
        public abstract Severity SeverityLevel { get; }
        public abstract void Verbose(string format, params object[] args);
        public abstract void Verbose2(string format, params object[] args);
        public abstract void Verbose3(string format, params object[] args);
        public abstract void Info(string format, params object[] args);
        public abstract void Error(int logCode, string message, Exception exception = null);
        public abstract void Warn(int logCode, string format, params object[] args);
        public abstract void Warn(int logCode, string message, Exception exception);
        public abstract void Info(int logCode, string format, params object[] args);
        public abstract void Verbose(int logCode, string format, params object[] args);
        public abstract void Verbose2(int logCode, string format, params object[] args);
        public abstract void Verbose3(int logCode, string format, params object[] args);

        void TrackDependency(string dependencyName, string commandName, DateTimeOffset, startTime, TimeSpan duration, bool success);
        void TrackEvent(string eventName, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null);
        void TrackMetric(string name, double value, IDictionary<string, string> properties = null);
        void TrackRequest(string name, DateTimeOffset startTime, TimeSpan duration, string responseCode);

        .... etc
}
```

I think something like this will keep the grain developer story simple. The rest will be between the Runtime and the telemetry providers. I don't have a good view into how the specific classes you listed will exactly fit there. It will probably become more clear in code. However, I did notice that you put RoleName and RoleInstance there. If you meant Azure PaaS terms, then I'm concerned that would be too Azure specific. If you meant it in a more general sense, then I think we better use different terms there to avoid confusion.

**Comment by galvesribeiro** [comment]
> I thought there was a general agreement that the client api would be an extension of the existing Orleans logging client. (See Orleans.Runtime.Logger).

OH, maybe I understood wrong. Well, current log API as @sergerbykov suggested would still be there to avoid broking something for a while, and that would be pushing logs to a LoggerProvider available and we would introduce new interface/API/property on Grain class so people that want use APM right now, can use it. Later on, we could depracate the old Logger API in flavor of the new one.

> The Orleans client api will likely need some cleanup before being extended, but generally what we come up with should be a (mostly) natural extension of the existing code.

I don't see why the suggested interfaces would be complicated. People want information. Just follow the current log API by extending it, is something pretty limited when you ask for people who are used to use any APM products.

> In this proposal, the client and provider interfaces are the same. Even if the provider and client interfaces end up looking exactly the same, which I don't think they will, they should be different interfaces, because they perform different responsibilities and need to be independent.

Ok, they can be different interfaces, but IMHO, they are there just as a proxy to receive the telemetry requests and push to any registered provider and that is why I think it must at least start with the same method signatures as the provider one.

> Things like SiloContext, GrainContext and the like should not be part of the client interfaces. Application developers shouldn't be concerned with this level of detail. The framework should fill in things like this and pass them through to the providers.

Yes I agree, those internal contexts are something filled by the internal implementation of the producer interface from data that come internal by the framework.

> I don't think the client interfaces can return tasks. Applications can't be slowed down by asynchronous calls to record performance data.

Yes! I'm sorry! It was a virus place in my head by @gabikliot for everything in Orleans has to return a Task to comply with the TaskScheduler hahahah jk! :) They should return void since it must be complete background work in order to push the telemetry without add any kind of latency in the request.


Anyway, I don't think the current client log APIs should be extended... I think it indeed must stay there for compatibility and change only its internal implementation to push data to a provider and add this new interface. We can even mark the current log [Deprecated] so people start moving to the new API.

**Comment by jason-bragg** [comment]
@sergeybykov I agree that Logger existing calls can probably stay as they are, I was kinda using this as an opportunity to reduce the surface areas of the interface.  Many of the Logger calls (Info, verbose, warning, overrides and such) can be extension functions built against an interface something like what I suggested.  The reduction of the surface area reduces cost of implementation and maintenance.

Having said that, none of the reasoning I just put forward is a goal for this change, so.. yeah.. feature creep..  :?

**Comment by galvesribeiro** [comment]
Ops! Race condition detected @sergeybykov ! heheheheh

Well my opinion remains. I think we must give the power to user to correlate events and classes like RequestTelemetry and (some) of the contexts, give the user the ability to make many telemetry events correlated. This is one of the richest features that APM tools use.

But if you guys think add only those methods to logger would be good? Ok, majority won :)

**Comment by sergeybykov** [comment]
@galvesribeiro Maybe we should do a Skype call to discuss this. I have a feeling there is a misunderstanding somewhere, potentially on my side. I'm failing to see the contradiction you seems to imply between extending the Logger API and the ability to correlate telemetry events.

**Comment by galvesribeiro** [comment]
I wish I could pull someone from Application Insights team to comment here. They made a terrific work merging many products on the market in order to bring a familiar interface for APM users and DevOps practitioners so that could be easy to move from virtually any APM tool from the market to AI.

Well, thinking again, you are right... why create a new interface and then move users from the old one to the new one? :( my bad...

As long as we have something like this:

```
Task TrackDependency(string dependencyName, string commandName, DateTimeOffset startTime, TimeSpan duration, bool success);
        Task TrackDependency(DependencyTelemetry telemetry);
        Task TrackEvent(EventTelemetry telemetry);
        Task TrackEvent(string eventName, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null);
        Task TrackMetric(MetricTelemetry telemetry);
        Task TrackMetric(string name, double value, IDictionary<string, string> properties = null);
        Task TrackRequest(RequestTelemetry request);
        Task TrackRequest(string name, DateTimeOffset startTime, TimeSpan duration, string responseCode, bool success);
        Task TrackTrace(TraceTelemetry telemetry);
```
... on current interface, there is no reason to create a new one... both of you are right. At the client side lets keep that this way and ofc, remove the internal context information from the requests.

The correlation on APM tools usually is made by some ID inside those context data so the user can inform this property when they create the context object.

So... Sorry about the confusion, let stick as you guys suggested :)

So, what else do we need to get a green light to start?

Thanks again for the support :)

**Comment by sergeybykov** [comment]
From my perspective, this is ready for the coding phase. I personally find it easier and more productive to iterate on a code proposal  that I can pull, change, and post back than on a .md. :-)

If you create an initial PR (I'd recommend to do that as soon as you have a bare minimum), we'll all be able to comment on the concrete code and submit PRs against your branch.

**Comment by jason-bragg** [comment]
Green light from my end as well.

Though, my suspicion is that this is going to be a significant change, so it might be valuable get a first pass worked out, with a minimal feature set, then add more capabilities in future iterations.  We don't need to do all of this in a single Omega-PR.

For instance, updating the existing Logger to pipe through to an ILoggerTelemetryProvider and testing that is not a trivial change in itself.

**Comment by galvesribeiro** [comment]
Thank you guys. The idea is exactly it. Get minimal commits on my fork and submit small PRs for that. Lets talking on Gitter an on the PR itself. I think we can close this issue.

Thanks

**Comment by sergeybykov** [comment]
Let's keep it open until the PRs are merged as we usually do. Gitter is too ephemeral for important things, and not everyone is on Gitter all the time. So let's put significant discussions on the PR or here.

**Comment by galvesribeiro** [comment]
Sure!

**Comment by centur** [comment]
Can I drop my 2 cents ? The discussion here is about 2 separate concepts Logging and Performance tracing. But isn't just 2 sides of one problem: 
As a Devops I want to see what's going on in my system in form of 
* performance metrics\counters (which APM tools can aggregate into bars and pies)
* custom events\ log messages
So in case of problem - I can identify bottlenecks OR bad input caused exceptions.

From a developer perspective this infrastructure must be backed by very fast intermediate storage on each platform thus it won't affect performance of my application. I think we know that a good example of such backing storage is ETW (not sure about Linux but there must be something). Also there is number of tools which can read from ETW and pass it down to other silos. 
The issue between Logger and Telemetry interfaces that Logger flattens everything down to a single string and for telemetry we want structured set of arguments that must be sent separately so APM can perform structured logging rather than simple text logging  (and use that structured approach later to show performance counters and so on). If logger will have extra method which accepts collection of telemetry details - it can be used for sending telemetry too. It just needs to change it's internal behaviour and instead of simple Trace.XXX use corresponding TraceSource (see [Trace vs TraceSource](http://stackoverflow.com/questions/1101174/difference-between-using-trace-and-tracesource) )
TraceSource supports separate parameters collection which stays structured and stored along with the flattened "message".

Then we can build our own custom LogConsumers that will receive calls and will send it to any APM system with the structured data.

![image](https://cloud.githubusercontent.com/assets/720822/11108758/e4f5d7ae-893d-11e5-9b8a-bf4fe0a130f4.png)
Here is an example of benefits from structured logging - 2 log entries for the same event in my local SEQ instance: 
first one (gray background) is what produced by my custom ILogConsumer writing directly into Serilog - as I'm not getting much data besides "message" - it's simply dumb text. 
second one is Trace listener that wired up directly to Trace class ( skipping Orleans custom logging and listening to what's being written directly to Trace) - not much more data but at least I can filter messages on Source context.

So the benefit of structured logging that extra details are not being lost when written storage. and later on I can filter on extra detail. If I can get datetime or performance counter as a structured item - I can filter and setup alerts on that. I think it's what all NewRelic, AppInsights  and other APM are doing. And I feel that Telemetry interface will be used for logging because it gives you an option to push structured data, not just text message.

**Comment by galvesribeiro** [comment]
Thanks for the comment but I respectfully disagree.

As previously mentioned, APM and Logging are 2 separate concepts.

Structured logging/logsource IMHO is better than just logs but still, logs…

NewRelic and other APM tools provider a terrific query language on top of the structured data. There is no (performatic) way to do that on flattened data. The level of detail of a log is so poor… It has just data, source, id, severity and mostly a message… The same for ETW and events pushed to event viewer and not to mention that logger is a pure technical component while APM is a technical AND business component.

It’s not a matter of which one is better but which one achieve the target here, have visibility of what is going on at a Orleans Cluster.

**Comment by centur** [comment]
Umm isn't it what I said - we need structured "logging" instead of flat one to record events with parameters instead of flat log strings ? As soon as we can have event object instead of string - we can decide where we want to send it - to APM\structured logging or flatten it to string logs. 

~~Also I pointed that ETW with TraceSource can be pretty reasonable decoupled storage for logs -  TraceSource supports object parameters which allows consumer to send structured data - similar to proposed Dictionary<string,string> and Dictionary<string,double>~~
I revised TraceSource signature and now feel that it's not the best idea as the object interpretation is fully on TraceListeners and it requires to build custom trace listener which will pull out object parameters before sending it to APM or log sinks. Operating with data objects directly is more reasonable approach

**Comment by galvesribeiro** [comment]
Understood. If you look at the beginning of the thread, I suggested a new Log API where objects can be passed along with the message text.

The majority want to keep the current log API at the client side as is, and just add the new APM/Telemetry methods. Both log and Telemetry APIs will forward to the logger/telemetry providers the request, and then push the data to the underlying tool.

When the first API get done, we could probably get more methods to the interface and add more ways to report log IMHO.

I’m just finishing some problems at my company so I can get back on this.

Thanks

**Comment by jason-bragg** [comment]
>"When the first API get done, we could probably get more methods to the interface and add more ways to report log"

Agreed.  Reading the recent posts I think we're mostly all on the same page.  The reasons for extending the existing logging infrastructure, rather than introducing a parallel one, are related to backwards compatibility and code organization.  I was not arguing that the existing logging calls are what engineers 'should' use, only that they are what engineers 'have been' using.  So for existing code (and there is lots of that) to benefit from these changes we need to improved the existing capabilities.

**Comment by galvesribeiro** [comment]
Sure!

**Comment by tsibelman** [comment]
@galvesribeiro  @sergeybykov I have a question, right now orleans is collecting various runtime statistics using varius counters using classes derived from ICounter interface, will you planing to refactor use of these counters with new metrics implementation ?

**Comment by galvesribeiro** [comment]
@tsibelman good question! The answer is yes! I'm going to implement the performance counters support on `OrleansTelemetry.PerformanceCounters` package and will replace all those ICounter implementations with calls to the telemetry APIs. This is actually required in order to move to coreCLR as in #368 since it does not support performance counters. Those metrics/counters will be received by all IMetricTelemetryConsumers and dispatched to the underlying technology. Actually, later on, all the current statistics implementations around Orleans will be pushed to it so we will have a common set of KPIs from the cluster, exposed by default by Orleans.

**Comment by tsibelman** [comment]
@galvesribeiro The counters in Orleans are actually a bit complicated, when you Increment some ICounter derived class you do not actualy comunicate with wind perf counters but with internal Orleans data structers. From these data structures statistics smaller subset is periodicly pushed into perf counters and in addition a full set is pushed into special providers implementing IConfigurableStatisticsPublisher interface (Orleans has two such implementation) that are responsible for publishing this data to SQL tables in bulk.  I know all this becouse I imlemented additional IConfigurableStatisticsPublisher provider that pushes the data into Metrics.NET counters, I don't like this aproach becouse for every internal Orleans counter I add identical Metrics.NET based counter. Idealy I would like to have an abilty to exchange orleans counters with Metrics.NET and not duplicate them.

So this way I existed about your initiative but I don't sure if you aware of this part of orleans and already has a plan for refactoring (for me it looks like it not a small change).  I can help you with some parts of this refactoring if you like.

**Comment by galvesribeiro** [comment]
@tsibelman yes I'm aware that there was a Statistic API in place. I'll dig further on that but the main idea is to remove the PerfCounter logic from inside orleans... The way it report those statistics could still be there (idk, need investigate) as long as it push the data thru the TelemetryConsumers and not to the perfcounters or Metrics.Net directly... 

Actually you will be able to create a Metrics.Net TelemetryConsumer when available.

**Comment by tsibelman** [comment]
So looks like it definatly moving into right directions this is great, thank you.

**Comment by eferron** [comment]
Have you looked at AppDynamics for this? I am working with a customer doing a large deployment on using the Orleans framework and I have had great success instrumenting their applications.

**Comment by sergeybykov** [comment]
@eferron Do you mean a telemetry consumer for AppDynamics?

Also, if we could list your customer on http://dotnet.github.io/orleans/Who-Is-Using-Orleans, that would be great. :-)

**Comment by eferron** [comment]
My customer is already on the list. We can chat more offline if you want
details, then determine what to make public where appropriate.

**Comment by sergeybykov** [comment]
Good.

If you already instrumented their app with AD, how much work do you think will it be to create an AD telemetry consumer?

**Comment by eferron** [comment]
I am reading thru the thread from what I can see so far this looks
achievable.

Let me comb thru this a bit and see if I can't have something before the
end of the week.

Ed

**Comment by sergeybykov** [comment]
:+1:  Let us know how we can help.

**Comment by jason-bragg** [comment]
Question about the intended usage of the telemetry apis:

Telemetry and performance counters are not exactly the same, is reasonable to have an implementation of IMetricTelemetryConsumer that writes to performance counters?

**Comment by galvesribeiro** [comment]
Hey @jason-bragg,

Telemetry is not exactly the same as Perf Counters indeed however, if you look at most used APM tools on the market (Azure Application Insights, NewRelic, DataDog, AppDynamics, and others), you will see that Per Counters are part of the Telemetry. With that said, the APM agents capture the data of perf counters and push to the underlying APM servers and make those counters metrics that will later be used to compose some KPI inside a Dashboard for instance.

Strictly talking about perf counters in Orleans, I see 2 ways that this can work:

1. `IPerformanceCounterTelemetryConsumer` (which can have a default implementation on Orleans as we have for File/Trace consumers) would have methods that will receive metrics and its underlying implementation will set/push values to a counter installed on that silo machine and later on, by any other tool, the data from that counter will be extracted and processed. This is the simplest implementation and will use ordinary counters on the underlying OS to record the data. (by underlying OS I mean not only Windows. Linux or OSX counters can be used for that as well when we move to .Net Core and support other OSses)
2. One implementation of `IPerformanceCounterTelemetryConsumer` could be made per APM tool which will record data the same way as on item 1 inside the underlying perf counter however, it could be configured to push/flush data from that counter regularly to the underlying APM server.

Does that make sense or do you have something else in mind? We could elaborate more on that implementation... I'm just explaining the simplistic high-level view of it... Remember that the interfaces for the Telemetry API wasn't carved on stone, so if we feel that perf counters deservers new methods we could do that. IMHO, I would actually use `IMetricTelemetryConsumer` and let the underlying APM tool make the aggregation and presentation of that metric instead of relying on perf counters. Talking about xplat, I would move away for perf counters and that was the main reason I suggested the implementation of this API in the first place...

**Comment by jason-bragg** [comment]
@galvesribeiro Thanks for your response.  I'll need to think on this a little.

It might help to explain what I'm up to.  While porting over the EventHubStreamProvider from 343 Industries, I removed all of the performance counters and logging.  I want to add this back in.  For the logging, I can use Orleans logs, but for the performance counters, I don't really see a good solution.  I could write the information out using the telemetry api, then write an IMetricTelemetryConsumer that converts these calls to perfcounters, but this is a bit hacky.

Additionally, the original code logged a lot of stuff, because it used perf counters and etw logs, which were both very light.  Logging that much stuff through Orleans logging abstraction in a code path that is very performance sensitive gives me pause.

I could introduce a specific logging interface to the provider, with a default implementation that uses Orleans logging and telemetry hooks.  Then the monitoring perfcouners and etw events from the original code could be moved inside the customer specific implementation of this interface.  This would work but be clumsy.

I could refactor Orleans application facing logging and telemetry code (Logger) to use interfaces and modify Logger to inherite from them, then have the application talk to these interfaces, which by default would use Logger and the Orleans stack.  Like the custom monitoring api, this would allow users of the provider to supply their own implementations of these interfaces that logged directly to etw, or perfcounters.

As I said.. still thinking :/

**Comment by galvesribeiro** [comment]
Just to notify the followers here. Performance counters are being removed from OrleansRuntime.dll 

More info #2122

**Comment by sergeybykov** [comment]
It seems to me that this issue has outlived its purposed already. We still have work to do (post 2.0.0) to improve the telemetry story. But I think that better be tracked with separate, more targeted, issues.

**Comment by galvesribeiro** [comment]
Sure. Lets revisit it after 2.0 release. The new issue is #3934

Thanks to everyone who contributed to the original idea. 😄

**Comment by sergeybykov** [comment]
Thanks a lot, indeed!
